platform/ios/src/MGLMapView.mm

```diff
@@ -4513,7 +4513,26 @@ - (void)notifyMapChange:(mbgl::MapChange)change
                 || self.userTrackingMode == MGLUserTrackingModeNone
                 || self.userTrackingState != MGLUserTrackingStateChanged)
             {
-                [self deselectAnnotation:self.selectedAnnotation animated:NO];
+                // Deselect annotation if it lies outside the viewport
```

Comment:
This is solving for something similar to `-[MGLMapView updateAnnotationViews]`. I wonder if checking if the -array `-[MGLMapView visibleAnnotations]` contains the selected annotation would be fast enough. If it is, it would simplify this implementation.